src/LiveryBridge.ts

```diff
@@ -0,0 +1,326 @@
+    }
+
+    return isReject;
+  }
+
+  static isResolveMessage<ValueType>(
+    object: LiveryMessage,
+  ): object is ResolveMessage<ValueType> {
+    return (
+      object.type === 'resolve' && !!(object as ResolveMessage<ValueType>).value
+    );
+  }
+
+  static propertyTypeCheck(
+    property: unknown,
+    propertyName: string,
+    messageName = '',
+    correctType: string,
+  ) {
+    if (typeof property !== correctType) {
+      throw new Error(
+        `Received ${messageName} message with ${propertyName} property type: ${typeof property}, should be ${correctType}`,
+      );
+    }
+  }
+
+  public sendCommand<ValueType, ArgType>(
+    name: string,
+    id = uuid(),
+    arg?: ArgType,
+  ): Promise<ValueType> {
+    return this.handshakePromise
+      .then(() => {
+        this.sendMessage('command', {
+          isLivery: true,
+          type: 'command',
+          name,
+          id,
+          arg,
+        });
+
+        return new Promise<ValueType>((resolve, reject) => {
+          this.deferredMap.set(id, { resolve, reject });
+        });
+      })
+      .catch((error) => Promise.reject(error));
+  }
+
+  public sendSubscribe<ValueType, ArgType>(
+    name: string,
+    listener: (value: ValueType) => void,
+    arg?: ArgType,
+  ) {
+    const id = uuid();
+    this.listenerMap.set(id, listener);
+    return this.sendCommand<string, ArgType>(name, id, arg);
+  }
+
+  // eslint-disable-next-line class-methods-use-this, @typescript-eslint/no-unused-vars -- Overridable method
+  protected handleCommand<ArgType>(
+    message: CommandMessage<ArgType>,
+  ): Promise<unknown> {
+    return Promise.reject(
+      new Error(
+        `Received command message with unsupported name: ${message.name}`,
+      ),
+    );
+  }
+
+  protected handleHandshake(message: HandshakeMessage) {
+    if (message.version !== this.version) {
+      this.sendReject(message.id, 'Versions do not correspond.');
+    } else if (message.id !== this.handshakeId) {
+      // Other side was last to send handshake
+      this.sendResolve(message.id, this.version);
+      const deferred = this.deferredMap.get(this.handshakeId);
+      if (deferred) {
+        deferred.resolve(this.version);
+        this.deferredMap.delete(this.handshakeId);
+      }
+      this.handshakeId = message.id;
+    }
+  }
+
+  protected handleLiveryMessage(message: LiveryMessage) {
+    if (LiveryBridge.isHandshakeMessage(message)) {
+      this.handleHandshake(message);
+    } else if (LiveryBridge.isResolveMessage(message)) {
+      this.handleResolveMessage(message);
+    } else if (LiveryBridge.isRejectMessage(message)) {
+      this.handleRejectMessage(message);
+    } else if (LiveryBridge.isCommandMessage(message)) {
+      this.handleCommandMessage(message);
+    } else if (LiveryBridge.isEventMessage(message)) {
+      this.handleEventMessage(message);
+    } else {
+      // should always be last
+      this.sendReject(message.id || 'unknown', 'Unknown Message');
```

Comment:
Should we send a reject here? Or just throw an error?

Reply:
Just throw an error; do try to make that error specific, e.g: `Unsupported message type: ${type}`